api/jsonrpc/server.go

```diff
@@ -175,58 +175,66 @@ func (j *JSONRPCServer) Execute(
 	defer span.End()
 
 	actionRegistry := j.vm.ActionRegistry()
-	action, err := (*actionRegistry).Unmarshal(codec.NewReader(args.Action, len(args.Action)))
-	if err != nil {
-		return fmt.Errorf("failed to unmashal action: %w", err)
+	actions := make([]chain.Action, 0)
+	for _, action := range args.Actions {
+		action, err := (*actionRegistry).Unmarshal(codec.NewReader(action, len(action)))
+		if err != nil {
+			return fmt.Errorf("failed to unmashal action: %w", err)
+		}
+		actions = append(actions, action)
 	}
 
 	now := time.Now().UnixMilli()
 
-	// Get expected state keys
-	stateKeysWithPermissions := action.StateKeys(args.Actor)
+	storage := make(map[string][]byte)
+	ts := tstate.New(1)
 
-	// flatten the map to a slice of keys
-	storageKeysToRead := make([][]byte, 0)
-	for key := range stateKeysWithPermissions {
-		storageKeysToRead = append(storageKeysToRead, []byte(key))
-	}
+	for _, action := range actions {
+		// Get expected state keys
+		stateKeysWithPermissions := action.StateKeys(args.Actor)
 
-	storage := make(map[string][]byte)
-	values, errs := j.vm.ReadState(ctx, storageKeysToRead)
-	for _, err := range errs {
-		if err != nil && !errors.Is(err, database.ErrNotFound) {
-			return fmt.Errorf("failed to read state: %w", err)
+		// flatten the map to a slice of keys
+		storageKeysToRead := make([][]byte, 0)
+		for key := range stateKeysWithPermissions {
+			storageKeysToRead = append(storageKeysToRead, []byte(key))
 		}
-	}
-	for i, value := range values {
-		if value == nil {
-			continue
+
+		values, errs := j.vm.ReadState(ctx, storageKeysToRead)
+		for _, err := range errs {
+			if err != nil && !errors.Is(err, database.ErrNotFound) {
+				return fmt.Errorf("failed to read state: %w", err)
+			}
+		}
+		for i, value := range values {
+			if value == nil {
+				continue
+			}
+			storage[string(storageKeysToRead[i])] = value
 		}
-		storage[string(storageKeysToRead[i])] = value
-	}
 
-	ts := tstate.New(1)
-	tsv := ts.NewView(stateKeysWithPermissions, storage)
-
-	output, err := action.Execute(
-		ctx,
-		j.vm.Rules(now),
-		tsv,
-		now,
-		args.Actor,
-		ids.Empty,
-	)
-	if err != nil {
-		reply.Error = fmt.Sprintf("failed to execute action: %s", err)
-		return nil
-	}
+		tsv := ts.NewView(stateKeysWithPermissions, storage)
+
+		output, err := action.Execute(
+			ctx,
+			j.vm.Rules(now),
+			tsv,
+			now,
+			args.Actor,
+			ids.Empty,
```

Comment:
This may cause issues for actions that rely on the `actionID` providing a unique ID.

Could we prevent this by using `CreateActionID(ids.Empty, actionIndex)` ? This should work because it will never conflict with a transaction that was actually issued onchain (there should never be a transaction with a zero value hash) and each action in the simulate API will get a unique `actionIndex` value. I think this should give us the best we can do here ie. on-chain execution results may differ with the actual `actionID`, but the `actionID` passed in to the simulation will always be unique.